File: packages/oxlint-plugin-react-doctor/src/plugin/utils/is-inside-statically-hidden-jsx-subtree.ts
import { getEffectiveStyleProperty } from "../rules/design/utils/get-effective-style-property.js";
import { getInlineStyleExpression } from "../rules/design/utils/get-inline-style-expression.js";
import { getStringFromClassNameAttr } from "../rules/design/utils/get-string-from-class-name-attr.js";
import { getStylePropertyStringValue } from "../rules/design/utils/get-style-property-string-value.js";
import type { EsTreeNode } from "./es-tree-node.js";
import type { EsTreeNodeOfType } from "./es-tree-node-of-type.js";
import { getAuthoritativeJsxAttribute } from "./get-authoritative-jsx-attribute.js";
import { getTailwindVisibilityAtBreakpoints } from "./get-tailwind-visibility-at-breakpoints.js";
import { isHiddenFromScreenReader } from "./is-hidden-from-screen-reader.js";
import { isNodeOfType } from "./is-node-of-type.js";
import type { RuleContext } from "./rule-context.js";

const isStaticallyHiddenOpeningElement = (
  openingElement: EsTreeNodeOfType<"JSXOpeningElement">,
  context: RuleContext,
): boolean => {
  if (isHiddenFromScreenReader(openingElement, context.settings)) return true;
  const styleAttribute = getAuthoritativeJsxAttribute(openingElement.attributes, "style");
  const styleExpression = styleAttribute
    ? getInlineStyleExpression(styleAttribute, context.scopes)
    : null;
  if (styleExpression) {
    const displayProperty = getEffectiveStyleProperty(styleExpression.properties, "display");
    if (displayProperty && getStylePropertyStringValue(displayProperty)?.toLowerCase() === "none") {
      return true;
    }
    const visibilityProperty = getEffectiveStyleProperty(styleExpression.properties, "visibility");
    const visibilityValue = visibilityProperty
      ? getStylePropertyStringValue(visibilityProperty)?.toLowerCase()
      : null;
    if (visibilityValue === "hidden" || visibilityValue === "collapse") return true;
  }
  const className = getStringFromClassNameAttr(openingElement);
  if (!className) return false;
  const visibilityAtBreakpoints = getTailwindVisibilityAtBreakpoints(className);
  return Boolean(
    visibilityAtBreakpoints && visibilityAtBreakpoints.every((isVisible) => !isVisible),
  );
};

export const isInsideStaticallyHiddenJsxSubtree = (
  node: EsTreeNode,
  context: RuleContext,
): boolean => {
  let currentNode: EsTreeNode | null | undefined = node;
  while (currentNode) {
    if (
      isNodeOfType(currentNode, "JSXElement") &&
      isStaticallyHiddenOpeningElement(currentNode.openingElement, context)
    ) {
      return true;
    }
    currentNode = currentNode.parent;
  }
  return false;
};

File: native/oxlint/rules/is-inside-statically-hidden-jsx-subtree.rs
fn is_inside_statically_hidden_jsx_subtree<'a>(
    node: &crate::AstNode<'a>,
    ctx: &crate::context::LintContext<'a>,
) -> bool {
    ctx.nodes().ancestors(node.id()).any(|ancestor| {
        let oxc_ast::AstKind::JSXElement(element) = ancestor.kind() else {
            return false;
        };
        is_statically_hidden_opening_element(&element.opening_element, ctx)
    })
}

fn is_statically_hidden_opening_element<'a>(
    opening_element: &'a oxc_ast::ast::JSXOpeningElement<'a>,
    ctx: &crate::context::LintContext<'a>,
) -> bool {
    if is_statically_hidden_from_screen_reader(opening_element, ctx) {
        return true;
    }
    let style_attribute = get_authoritative_jsx_attribute(opening_element, "style", true);
    if let Some(style_expression) =
        style_attribute
            .and_then(|attribute| get_inline_style_object_expression_with_aliases(attribute, ctx))
    {
        if get_effective_static_style_property(style_expression, "display")
            .and_then(get_object_property_string_value)
            .is_some_and(|value| value.eq_ignore_ascii_case("none"))
        {
            return true;
        }
        if get_effective_static_style_property(style_expression, "visibility")
            .and_then(get_object_property_string_value)
            .is_some_and(|value| {
                value.eq_ignore_ascii_case("hidden") || value.eq_ignore_ascii_case("collapse")
            })
        {
            return true;
        }
    }
    let Some(class_name) = get_static_class_name(opening_element) else {
        return false;
    };
    get_tailwind_visibility_at_breakpoints(class_name)
        .is_some_and(|visibility| visibility.iter().all(|is_visible| !is_visible))
}
